### PR TITLE
Match analyzer and linker requires methods

### PR DESCRIPTION
**Pull request by tlakollo**
Match analyzer and linker requires methods to be more similar according to their behavior
Sync diagnostic Id in NativeAOT with the information in linker

### CODE REVIEW

File: src/ILLink.Shared/DiagnosticId.cs [hunk 1]
```diff
@@ -187,7 +187,11 @@ public enum DiagnosticId
 		// Dynamic code diagnostic ids.
 		RequiresDynamicCode = 3050,
 		RequiresDynamicCodeAttributeMismatch = 3051,
-		RequiresDynamicCodeOnStaticConstructor = 3052
+		RequiresDynamicCodeOnStaticConstructor = 3052,
+		COMInteropNotSupportedInFullAOT = 3052,
```
Comment (tlakollo) [inline src/ILLink.Shared/DiagnosticId.cs:191]: Notice that there is a collision on the code IL3052, I don't have a strong preference on which one should remain IL3052 and which one should change, but one should be changed.

Comment (vitek-karas) [reply src/ILLink.Shared/DiagnosticId.cs:191]: Did we ship one of them in .NET 6? If only one, then we should keep that one as is and change the new one.
If not... I guess just pick one...

File: src/ILLink.RoslynAnalyzer/RequiresAnalyzerBase.cs [hunk 2]
```diff
@@ -183,7 +183,7 @@ public override void Initialize (AnalysisContext context)
 							if (instanceCtor.Arity > 0)
 								continue;
 
-							if (instanceCtor.TargetHasRequiresAttribute (RequiresAttributeName, out var requiresAttribute) &&
+							if (instanceCtor.DoesMemberRequires (RequiresAttributeName, out var requiresAttribute) &&
```
Comment (vitek-karas) [inline src/ILLink.RoslynAnalyzer/RequiresAnalyzerBase.cs:186]: Nit - not for this PR: The grammar of the new name is really weird - "Does member requires" feels wrong, I think it should read "Does member require <something>". I know it's how we have it everywhere, so not for this change, but maybe something to consider one day :-)

**Review by vitek-karas — APPROVED** [review]

File: src/ILLink.Shared/DiagnosticId.cs (hunk 1 above)
Comment (tlakollo) [reply src/ILLink.Shared/DiagnosticId.cs:191]: COMInteropNotSupportedInFullAOT in NativeAOT seems to exist even before runtimelab project existed, so probably I will leave that as 3052 even if we haven't shipped NativeAOT yet

**Review by vitek-karas — APPROVED** [review]

**Pull request merged**